### ISSUE TITLE
operation use-gcs-blobstore-service-account.yml is not compatiable with cf-deployment.yml

### ISSUE THREAD

**Issue by alexwo** [cloudfoundry/cf-deployment#429]
Running Bosh interpolate with use-gcs-blobstore-service-account.yml  does not succeed.

I have bumped into the following list of issues:

1.Above operation will try to modify not existing properties in cf-deployment.yml.
There are no  "fog_connection" or  &fog_config properties defined in the cf-deployment.yml. (only blobstore-properties)

2.The dir keys properties var mappings below are also missing on the operation ->use-gcs-blobstore-service-account.yml but appear in the example file for this operation.

```
blobstore-service-account.yml` defination:
app_package_directory_key: example-apps-dir
buildpack_directory_key: example-buildpacks-dir
droplet_directory_key: example-droplets-dir
resource_directory_key: example-resources-dir
are completely missing in the operation file 
```
example:
https://github.com/cloudfoundry/cf-deployment/blob/master/operations/example-vars-files/vars-use-gcs-blobstore-service-account.yml

3.In the alternative use-gcs-blobstore.yml operation, there are "remove" sections which are missing in  -gcs-blobstore-service-account.yml but seem to be required for GCP as:

```
# remove internal blobstore including its variables
- type: remove
  path: /instance_groups/name=singleton-blobstore

- type: remove
  path: /variables/name=blobstore_admin_users_password

- type: remove
  path: /variables/name=blobstore_secure_link_secret

- type: remove
  path: /variables/name=blobstore_tls
```
https://github.com/cloudfoundry/cf-deployment/blob/master/operations/use-gcs-blobstore.yml

Your feedback is very much appreciated.

thanks,
Alex

**Comment by johnsonj** [comment]
Squinting at the ops files, I wonder if it expects the `use-gcs-blobstore.yml` and `use-gcs-blobstore-service-account.yml` be applied in that order. Can @selzoc @tylerschultz confirm?

I'd expect those two to be distinct operations and if possible rename `use-gcs-blobstore-yml` to `use-gcs-blobstore-hmac.yml` (and maybe add deprecated).

**Comment by selzoc** [comment]
Please refer to the [https://github.com/cloudfoundry/cf-deployment/blob/master/operations/README.md](README) which is where we document opsfile requirements.  From the entry for `use-gcs-blobstore-service-account.yml`:
```
Requires use-gcs-blobstore.yml. Introduces new variables for gcp service account credentials.
```

Why would you want these to be distinct operations?  Originally the opsfile only supported the AWS-style key id and secret, and later when support was added to CAPI for the json api, we added the service account opsfile.

**Comment by johnsonj** [comment]
Thanks @selzoc! 

I'm biased on this because from my perspective the S3-compatibility mode (HMAC keys) and Native GCS (service account) are distinctly different things.

- The AWS-style key id and secret for GCS should be deprecated. We did the work to support service accounts because they provide a better security model and performance than our S3 compatibility mode. This is probably a separate conversation. 
- The UX of layering seems weird. I can apply just `use-gcs-blobstore.yml` and use `vars-use-gcs-blobstore.yml` OR I can apply `use-gcs-blobstore.yml, use-gcs-blobstore-service-account.yml` and use `vars-use-gcs-blobstore-service-account.yml`. This bug as an example of how that's confusing and my concern is it's misleading because the two use such different forms of authentication. 
- In my opinion `use-gcs-blobstore-service-account.yml` is as independant from `use-gcs-blobstore.yml` as `use-azure-storage-blobstore.yml`

**Comment by selzoc** [comment]
Cool, thanks for the context!  @dsabeti may want to weigh in on the independence of opsfiles - we have found that there is a delicate balance between an opsfile for every single permutation, or being able to compose them to suit your needs.

**Comment by dsabeti** [comment]
Thanks for chiming in @johnsonj and @selzoc!

 I think this comment really jolted my understanding of the problem:
> In my opinion use-gcs-blobstore-service-account.yml is as independant from use-gcs-blobstore.yml as use-azure-storage-blobstore.yml

At first glance, it seemed incorrect. But insofar as all of ops-files for using external blobstores reconfigure the CC almost identically -- the only difference are the fog connection properties -- the GCS-with-access-keys and GCS-with-service-account ops-files are in fact as independent from each other as any other blobstore-related ops-files.

I'm starting to wonder if cf-deployment should reference fog more directly. For example, instead support a generic `use-fog-blobstore-connection.yml`, which would transparently pass the fog properties to the CC and work for _any_ blobstore that fog also supports:
```
bosh deploy cf-deployment.yml \
  -o operations/use-fog-blobstore-connection.yml \
  -v fog_hash="{...}"
```

Another idea (which I think I like better) would be to have an ops-file to switch to fog, and then a set of ops-files that help to configure the fog hash depending on your desired blobstore:
```
bosh deploy cf-deployment.yml \
  -o operations/use-fog-blobstore-connection.yml \
  -o operations/use-gcs-with-service-account.yml \
  -v gcs_service_account_json_key=... \
  -v gcs_project=...
```
The advantage of this latter approach would be that we could continue to give feedback to operators _at deploy time_ if they've misconfigured their fog connections.

Any thoughts on those approaches?

cc @heyjoshcollins @staylor14

**Comment by dsabeti** [comment]
Ok, sounds good. We probably won't get to it for a few weeks, but I'd like to see this change happen. Given that it's a breaking change, we'll probably batch it up with a few other breaking changes.

**Comment by firasnajjar** [comment]
@dsabeti are there anytimlines for this change?

**Comment by dsabeti** [comment]
@firasnajjar Not yet. According to our tracker, it's still several weeks out, but it's not exactly high fidelity right now.

**Comment by ninja-at-work** [comment]
Hi @dsabeti , 

When can we expect this change to be done? Or at which version it could be available (v1.26.0? or later?) if the tracker is not accurate at this moment.

Thanks

**Comment by dsabeti** [comment]
Hey @ninja-at-work, honestly, not for some time as we have a lot of other things to work on. Still, this change shouldn't block anything as there's a workaround (albeit an unsatisfactory one). If you want to use a service account key, just apply both ops-files: `-o use-gcs-blobstore.yml -o use-gcs-blobstore-service-account.yml`.

**Comment by ishustava** [comment]
Hey @alexwo @johnsonj  @firasnajjar @ninja-at-work 

We shipped this change in [cf-deployment 2.0](https://github.com/cloudfoundry/cf-deployment/releases/tag/v2.0.0). I'm going to close this issue, but please feel free to open another one if you see any problems with the new approach.
